packages/engine/lib/error/src/frame.rs

```diff
@@ -0,0 +1,275 @@
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.0, fmt)
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T: fmt::Debug> fmt::Debug for StdError<T> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&self.0, fmt)
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T: std::error::Error> Provider for StdError<T> {
+    fn provide<'p>(&'p self, req: &mut Requisition<'p, '_>) {
+        #[cfg(feature = "backtrace")]
+        if let Some(backtrace) = self.0.backtrace() {
+            req.provide_ref(backtrace);
+        }
+    }
+}
+
+impl VTable {
+    /// Drops the `frame`
+    ///
+    /// # Safety
+    ///
+    /// - Layout of `*frame` must match `FrameRepr<E>`.
+    unsafe fn object_drop<E>(frame: Box<FrameRepr>) {
+        // Attach E's native vtable onto the pointer to self._error
+        let unerased = Box::from_raw(Box::into_raw(frame).cast::<FrameRepr<E>>());
+        drop(unerased);
+    }
+
+    /// Unerase error as `&dyn Error`
+    ///
+    /// # Safety
+    ///
+    /// - Layout of `Self` must match `FrameRepr<E>`.
+    unsafe fn object_ref<E: Context>(frame: &FrameRepr) -> &dyn Context {
+        // Attach E's native vtable onto the pointer to self._error
+        let unerased = (frame as *const FrameRepr).cast::<FrameRepr<E>>();
+        // inside of vtable it's allowed to access `_error`
+        #[allow(clippy::used_underscore_binding)]
+        &(*(unerased))._error
+    }
+
+    /// Downcasts error to `target`
+    ///
+    /// # Safety
+    ///
+    /// - Layout of `Self` must match `FrameRepr<E>`.
+    unsafe fn object_downcast<E: Context>(
+        frame: &FrameRepr,
+        target: TypeId,
+    ) -> Option<NonNull<()>> {
+        if TypeId::of::<E>() == target {
+            // Attach E's native vtable onto the pointer to self._error
+            let unerased = (frame as *const FrameRepr).cast::<FrameRepr<E>>();
+            // inside of vtable it's allowed to access `_error`
+            #[allow(clippy::used_underscore_binding)]
+            let addr = &(*(unerased))._error as *const E as *mut ();
+            Some(NonNull::new_unchecked(addr))
+        } else {
+            None
+        }
+    }
+}
+
+#[repr(C)]
+#[allow(clippy::module_name_repetitions)]
+pub struct FrameRepr<E = ()> {
+    vtable: &'static VTable,
+    // Must not be used directly, only through vtable
+    _error: E,
+}
+
+impl<E> FrameRepr<E> {
+    pub fn new(error: E) -> Box<FrameRepr>
+    where
+        E: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        let unerased = Box::new(Self {
+            vtable: &VTable {
+                object_drop: VTable::object_drop::<E>,
+                object_ref: VTable::object_ref::<E>,
+                object_downcast: VTable::object_downcast::<E>,
+            },
+            _error: error,
+        });
+        unsafe { Box::from_raw(Box::into_raw(unerased).cast()) }
```

Comment:
I would add a comment that transmuting `error` to a byte slice instead of the fake cast to `()` doesn't work for complicated reasons